### ISSUE TITLE
100% test coverage

### ISSUE THREAD

**Issue by piranna** [NodeOS/NodeOS#75]
If some software really needs to be reliable, that's the OS. We need to document and tests all the processes, and we must work to achieve a 100% testing coverage to be sure there's no hidden bugs.

**Comment by piranna** [comment]
I've just developed [coverdeeps](https://github.com/piranna/coverdeeps), a small tool that check a module dependencies and give a global tests coverage percentage instead of just only the own module, because having reliable dependencies are as important (or more!) than the module itself. It still lacks a lot of features, only support modules published on npm registry (not local ones or published only on a repository) and tests coverage info from coveralls.io so it's really pesimistic about the projects tests coverage, but at least it's a beginning :-)

And as usual, pull-requests are welcome :-D

**Comment by lite20** [comment]
This tool is fun to play with! I've been running 'coverdeeps' on everything for the last 20 minutes

**Comment by piranna** [comment]
And what's your impression? :-D

**Comment by lite20** [comment]
I get the impression that I have a lot of tests to write for a lot ;-;

**Comment by piranna** [comment]
> I get the impression that I have a lot of tests to write for a lot ;-;

I think this is a good thing, isn't it? ;-)

**Comment by lite20** [comment]
hahaha nope! Because that means I have a lot of work, but it will be nice once that number raises haha!

**Comment by piranna** [comment]
First of all, make coverdeeps more flexible... and get it to achieve 100%
;-)

**Comment by luii** [comment]
Can we make a list of all things we need to test because i wann to help

**Comment by piranna** [comment]
Coverdeeps should help on this by showing the tests coverage of each
module, but I think we can start with itself. You know, eat our own dog
food :-)

**Comment by lite20** [comment]
For a list of tasks to complete, there is the roadmap. You can find a link
to it on the Wiki

**Comment by lite20** [comment]
Just for clarification, does NodeOS support native modules (compiled
modules)

**Comment by piranna** [comment]
Not to install them with npm from inside NodeOS itself because we don't have yet a c compiler (probably nodeos-cross-toolchain would be reused for this), but you can add them on build time.

**Comment by Compynerd255** [comment]
Quick question: How are you defining coverage? If you're purely going by lines of code (which is what it appears that coverdeeps is doing), then you aren't necessarily covering all expected behaviors, since you have varying branches and whatnot. Are you accounting for branches, including branches implicit in operations such as division (which throw errors when you try to divide by zero)?

**Comment by piranna** [comment]
Good question. At this moment I only trust on what coveralls.io does, so if
it only count lines (I think it do), then that's what we have. Coverdeeps
is really pesimistic about percentage, that's something good since if we
get some better checkers the value given can only decrease... so we have
more things to fix :-) We could expand it to use bithound.io or other
metrics, too...

**Comment by Compynerd255** [comment]
> Good question. At this moment I only trust on what coveralls.io does, so if it only count lines (I think it do), then that's what we have. Coverdeeps is really pesimistic about percentage, that's something good since if we get some better checkers the value given can only decrease... so we have more things to fix :-) We could expand it to use bithound.io or other metrics, too...

Those things could work - having a metric like this is definitely a good first step to have. 

If you're trying to go for a perfectly bug free OS, however, then something that you might want to consider is actually going in and formally verifying the OS components (i.e. sitting down with the JS source and mathematically proving that each function actually does exactly what it's supposed to do). I believe there are tools that help automate this process (for instance, there's one called Coq that works for Haskell). It is an expensive process and not even possible in some cases (not to mention that Chrome V8 has bugs, so your proof can only be as correct as the V8 engine is), but if you can do it, then you never have to worry about bugs for that module ever again as long as it keeps the same feature set.

And if code is so convoluted that you can't figure out what it's supposed to do, that might be an indication to use a different module.

**Comment by piranna** [comment]
> If you're trying to go for a perfectly bug free OS, however, then something that you might want to consider is actually going in and formally verifying the OS components (i.e. sitting down with the JS source and mathematically proving that each function actually does exactly what it's supposed to do). I believe there are tools that help automate this process (for instance, there's one called Coq that works for Haskell). It is an expensive process and not even possible in some cases (not to mention that Chrome V8 has bugs, so your proof can only be as correct as the V8 engine is), but if you can do it, then you never have to worry about bugs for that module ever again as long as it keeps the same feature set.

That's definitely dificult to achieve, since we are using both Linux and Node.js (and as you've said, this one use v8), and the two are really big projects, and as far as I've seen there's no way to get tests coverage statistics for them, imagine to get a mathematic audit of the code... The best we could achieve is to claim that there's a perfect score of 100% on NodeOS core modules and dependencies, this itself would bring a HUGE confidence on the system, isn't it? :-)


> And if code is so convoluted that you can't figure out what it's supposed to do, that might be an indication to use a different module.

Or split it and make it simpler :-)

**Comment by Compynerd255** [comment]
> The best we could achieve is to claim that there's a perfect score of 100% on NodeOS core modules and dependencies, this itself would bring a HUGE confidence on the system, isn't it? :-)

Yes, that we could. I'm just saying that if we can prove that the JS that drives NodeOS core is correct, then we know that any bugs in NodeOS that we do find are V8's fault or the Linux kernel's fault, not NodeOS's fault. And then our JS code can serve as a reproducible example that we can draw up really quickly for the bug report.

**Comment by piranna** [comment]
That's exactly what I was telling you.

**Comment by Compynerd255** [comment]
> That's exactly what I was telling you.

OK, that makes sense. :)
